File: 2856-count-complete-subarrays-in-an-array/count-complete-subarrays-in-an-array.py
from collections import Counter
class Solution:
    def countCompleteSubarrays(self, nums: List[int]) -> int:       
        n = len(nums)
        # 1. Compute total distinct count
        distinct_total = len(set(nums))
        
        freq = Counter()
        have = 0
        ans = 0
        
        l = 0
        # 2. Slide r from 0 to n-1
        for r in range(n):
            # Add nums[r]
            freq[nums[r]] += 1
            if freq[nums[r]] == 1:
                have += 1
            
            # 3. While window [l..r] is complete, count and shrink from left
            while have == distinct_total:
                # All subarrays from l to any end ≥ r are complete
                ans += (n - r)
                
                # Remove nums[l] and advance l
                freq[nums[l]] -= 1
                if freq[nums[l]] == 0:
                    have -= 1
                l += 1
        
        return ans

File: 2856-count-complete-subarrays-in-an-array/count-complete-subarrays-in-an-array.cpp
class Solution {
public:
    int countCompleteSubarrays(vector<int>& nums) {
        int n = nums.size();
        // 1. Total number of distinct values in nums
        unordered_set<int> distinct_set(nums.begin(), nums.end());
        int need = distinct_set.size();

        unordered_map<int, int> freq;
        int have = 0;
        long long ans = 0;
        int l = 0;

        // 2. Slide r from 0 to n-1
        for (int r = 0; r < n; ++r) {
            // Add nums[r]
            if (++freq[nums[r]] == 1) {
                // first time this value appears in window
                ++have;
            }

            // 3. While window [l..r] is complete, count and shrink from left
            while (have == need) {
                // All subarrays starting at l with end >= r are complete
                ans += (n - r);

                // Remove nums[l]
                if (--freq[nums[l]] == 0) {
                    // lost that distinct value
                    --have;
                }
                ++l;
            }
        }

        return ans;
    }
};